docs/KustomizeBestPractices.md

```diff
@@ -1,22 +1,133 @@
+Using a content hash ensures that rolling updates are triggered if the config map is changed.
+
+**Deprecated patterns**
+
+* vars should no longer be used to do command line substitution see [bit.ly/kf_kustomize_v3](https://docs.google.com/document/d/1jBayuR5YvhuGcIVAgB1F_q4NrlzUryZPyI9lCdkFTcw/edit?pli=1#heading=h.ychbuvw81fj7)
+
+## Eschew vars
+
+As noted in [kubernetes-sigs/kustomize#2052](https://github.com/kubernetes-sigs/kustomize/issues/2052) vars have a lot of downsides.
+For Kubeflow in particular vars have made it difficult to compose kustomize packages because they need to be unique globally ([kubeflow/manifests#1007](https://github.com/kubeflow/manifests/issues/1007)).
+
+Vars should be used sparingly. Below are some guidance on acceptable use cases.
+
+
+### Internal subsitution of fields Kustomize isn't aware of
+
+One ok use case for vars is getting kustomize to subsitute a value into a field kustomize wouldn't normally do substitution into. 
+This often happens with CRDs. For example, consider the virtual service below from [jupyter-web-app](https://github.com/kubeflow/manifests/blob/master/jupyter/jupyter-web-app/overlays/istio/virtual-service.yaml).
+
+```
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: jupyter-web-app
+spec:
+  gateways:
+  - kubeflow-gateway
+  hosts:
+  - '*'
+  http:
+  - ...
+    route:
+    - destination:
+        host: jupyter-web-app-service.$(jupyter-web-app-namespace).svc.$(clusterDomain)
+        port:
+          number: 80
+```
+
+We would like kustomize to substitute namespace into the destination host. We do this by
+
+1. Defining a [vars](https://github.com/kubeflow/manifests/blob/393ec700e7834ca69a0832ec01ea2ecd90fb5bc4/jupyter/jupyter-web-app/base/kustomization.yaml#L63) to get the value for namespace
+1. Defining a [custom configuration](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/transformerconfigs/README.md#customizing-transformer-configurations) so that the vars will be substituted into the virtual service host.
+
+This use of vars is acceptable because the var is internal to the kustomize package and can be given a unique enough name to prevent
+conflicts when the package is composed with other applications.
+
+### Global substitution
+
+One of the most problematic use cases for vars in Kubeflow today is substituting a user supplied value into multiple applications.
+
+Currently we only have one use case which is substituting in cluster domain into virtual services ([ref](https://docs.google.com/document/d/1jBayuR5YvhuGcIVAgB1F_q4NrlzUryZPyI9lCdkFTcw/edit#heading=h.vyq4iltpirga)).
+
+We would ultimately like to get rid of the use of vars in these cases but have not settled on precise solutions. Some possible options are
```

Comment:
IMO stating that you should not use global substitution except where there is no other option is enough in the best practices, no need to delve into possible solutions. Maybe just make a overarching issue where discussion can be had and just link there "if you want more info"?

Its fine as is. just my preference.

Reply:
I think I'd like to nudge people towards using kpt setters. I guess we could just as easily link to an issue with recommended patterns and describe uses there.

Reply:
Thanks @Klaven this is a good suggestion. I thought about following your suggestion and just linking to the bug kubeflow/manifests#1007. The issue is getting quite long though so then I thought it would be valuable to kind of summarize it here.  I also want to nudge people to start trying/using `kpt setters`.